quint/src/repl.ts

```diff
@@ -478,41 +487,65 @@ function loadShadowVars(state: ReplState, context: CompilationContext): void {
 
 // Declarations that are overloaded by the simulator.
 // In the future, we will declare them in a separate module.
-const simulatorBuiltins = `val ${lastTraceName} = [];
-def q::test(q::nruns, q::nsteps, q::init, q::next, q::inv) = false;
-def q::testOnce(q::nsteps, q::init, q::next, q::inv) = false;
-`
+function simulatorBuiltins(compilationState: CompilationState): FlatDef[] {
+  return [
+    parseDefOrThrow(compilationState, `val ${lastTraceName} = []`),
+    parseDefOrThrow(compilationState, `def q::test = (q::nruns, q::nsteps, q::init, q::next, q::inv) => false`),
+    parseDefOrThrow(compilationState, `def q::testOnce = (q::nsteps, q::init, q::next, q::inv) => false`),
+  ]
+}
 
-// try to evaluate the expression in a string and print it, if successful
-function tryEval(out: writer, state: ReplState, newInput: string): boolean {
-  // output errors to the console in red
-  function printErrors(moduleText: string, context: CompilationContext) {
-    printErrorMessages(out, 'syntax error', newInput, moduleText, context.syntaxErrors)
-    printErrorMessages(out, 'static analysis error', newInput, moduleText, context.analysisErrors, chalk.yellow)
-    printErrorMessages(out, 'compile error', moduleText, newInput, context.compileErrors)
-    out('\n') // be nice to external programs
+function parseDefOrThrow(compilationState: CompilationState, text: string): FlatDef {
+  const result = parseExpressionOrUnit(text, '<builtins>', compilationState.idGen, compilationState.sourceMap)
+  if (result.kind === 'toplevel' && isFlat(result.def)) {
```

Comment:
Does it mean that we would not be able to write nested definitions in REPL anymore?

Reply:
Not at all, this is only used for parsing the builtins. Also, `toplevel` here is just the kind returned by `parseExpressionOrUnit` when it finds a unit, so that would also work for let-ins (assuming that's what you mean by nested definitions).